bootloaders/eboot/eboot.ld

```diff
@@ -154,12 +154,24 @@ SECTIONS
     *(.gnu.linkonce.b.*)
     . = ALIGN (8);
     _bss_end = ABSOLUTE(.);
-    /* CRC stored in last 8 bytes */
-    ASSERT((. < 4096 - 8), "Error: No space for CRC in bootloader sector.");
-    . = _stext + 4096 - 8;
-    _crc_size = .;
-    . = . + 4;
-    _crc_val = .;
+    _free_space = 4096 - 17 - (. - _stext);
+/*
+The boot loader checksum must be before the CRC, which is written by elf2bin.py.
+This leaves 16 bytes after the checksum for the CRC placed at the end of the
+4096-byte sector. */
+    _cs_here = (ALIGN((. + 1), 16) == ALIGN(16)) ? (ALIGN(16) - 1) : (. + 0x0F);
```

Comment:
Sorry, but I don't follow what exactly is here.  The ROM segment checksum is placed outside of the ELF address space.  In the flash it follows the first sector's data (always 4096 bytes).  What is _cs_here representing, and who is setting/using it?

Reply:
Sorry, my experience with the gnu linker is lacking. The goal of this ugly expression was to calculate a location for the CS that landed just outside the `.text` section at the end of a 16 bit aligned. Its purpose was for a human cross-check. I used `hexdump` to view the .bin file created by `elf2bin.py`. I verified that CS and CRC were in the right places.

> In the flash it follows the first sector's data (always 4096 bytes).

Sorry, I am not sure I am parsing that correctly. So I maybe miss answering your question.
I could not find any detailed documentation on what the Boot ROM expected. So I resorted to the empirical method. From that, the Boot ROM expects a checksum at the end of the 16-byte alignment immediately following the last segment's payload. Since the CRC is at the end of the 4096-byte sector, the CS needs to at least proceed it. Otherwise when the CRC is written the CS will be in error.

Reply:
Got it.  Didn't understand the reasoning before.  This seems fine and fixes something I overlooked.